secboot/secboot_tpm.go

```diff
@@ -318,35 +318,81 @@ var fdeRevealKeyRuntimeMax = "2m"
 // overridden in tests
 var fdeRevealKeyCommandExtra []string
 
-// fdeRevealKeyCommand returns a *exec.Cmd that is suitable to run
-// fde-reveal-key using systemd-run
-func fdeRevealKeyCommand() *exec.Cmd {
+// runFDERevealKeyCommand returns the output of fde-reveal-key run
+// with systemd.
+//
+// Note that systemd-run in the initrd can only talk to the private
+// systemd bus so this cannot use "--pipe" or "--wait", see
+// https://github.com/snapcore/core-initrd/issues/13
+func runFDERevealKeyCommand(stdin []byte) (output []byte, err error) {
+	runDir := filepath.Join(dirs.GlobalRootDir, "/run/fde-reveal-key")
+	if err := os.MkdirAll(runDir, 0700); err != nil {
+		return nil, fmt.Errorf("cannot create tmp dir for fde-reveal-key: %v", err)
+	}
+	err = ioutil.WriteFile(filepath.Join(runDir, "fde-reveal-key.stdin"), stdin, 0600)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create stdin for fde-reveal-key: %v", err)
+	}
+
 	// TODO: put this into a new "systemd/run" package
 	cmd := exec.Command(
 		"systemd-run",
-		"--pipe", "--same-dir", "--wait", "--collect",
+		"--collect",
 		"--service-type=exec",
 		"--quiet",
 		// ensure we get some result from the hook within a
 		// reasonable timeout and output from systemd if
 		// things go wrong
 		fmt.Sprintf("--property=RuntimeMaxSec=%s", fdeRevealKeyRuntimeMax),
-		// this ensures we get useful output for e.g. segfaults
-		`--property=ExecStopPost=/bin/sh -c 'if [ "$EXIT_STATUS" != 0 ]; then echo "service result: $SERVICE_RESULT" 1>&2; fi'`,
 		// Do not allow mounting, this ensures hooks in initrd
 		// can not mess around with ubuntu-data.
 		//
 		// Note that this is not about perfect confinement, more about
 		// making sure that people using the hook know that we do not
 		// want them to mess around outside of just providing unseal.
 		"--property=SystemCallFilter=~@mount",
+		// WORKAROUNDS
+		// workaround the lack of "--pipe"
+		fmt.Sprintf("--property=StandardInput=file:%s/fde-reveal-key.stdin", runDir),
+		fmt.Sprintf("--property=StandardOutput=file:%s/fde-reveal-key.stdout", runDir),
+		fmt.Sprintf("--property=StandardError=file:%s/fde-reveal-key.stderr", runDir),
+		fmt.Sprintf(`--property=ExecStopPost=/bin/sh -c 'if [ "$EXIT_STATUS" = 0 ]; then touch %[1]s/fde-reveal-key.success; else echo "service result: $SERVICE_RESULT" >%[1]s/fde-reveal-key.failed; fi'`, runDir),
```

Comment:
could you add the comment about "this ensures we get useful output for e.g. segfaults back here? I think it's useful

Reply:
I added this back